File: src/org/blitzortung/android/data/beans/Station.java
package org.blitzortung.android.data.beans;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.TimeZone;

import org.json.JSONArray;
import org.json.JSONException;

public class Station {

	public enum State {
		ON, DELAYED, OFF
	}

	private String name;

	private float longitude;

	private float latitude;

	Date offlineSince;

	State state;

	public Station(JSONArray jsonArray) {
		TimeZone tz = TimeZone.getTimeZone("UTC");
		SimpleDateFormat formatter = new SimpleDateFormat("yyyyMMdd'T'HH:mm:ss");
		formatter.setTimeZone(tz);

		try {
			name = jsonArray.getString(1);
			longitude = (float) jsonArray.getDouble(3);
			latitude = (float) jsonArray.getDouble(4);
			if (jsonArray.length() >= 6) {

				offlineSince = formatter.parse(jsonArray.getString(5));
				
				Date now = new GregorianCalendar().getTime();
				
				int minutesAgo = (int) (now.getTime() - offlineSince.getTime()) / 1000 / 60;
				
				if (minutesAgo > 24*60)
					state = State.OFF;
				else if (minutesAgo > 15)
					state = State.DELAYED;
				else
					state = State.ON;

				
			} else {
				state = State.ON;
			}
		} catch (ParseException e) {
			throw new RuntimeException("error parsing station data");
		} catch (JSONException e) {
			throw new RuntimeException("error with JSON format while parsing station data");
		}
	}

	public String getName() {
		return name;
	}

	public float getLongitude() {
		return longitude;
	}

	public float getLatitude() {
		return latitude;
	}

	public Date getOfflineSince() {
		return offlineSince;
	}
	
	public State getState() {
		return state;
	}
}

File: app/src/main/java/org/blitzortung/android/data/beans/Station.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.data.beans

import java.io.Serializable

class Station(
    override val longitude: Double,
    override val latitude: Double,
    val name: String,
    val offlineSince: Long
) : Location, Serializable {

    val state: State
        get() {
            return if (offlineSince == OFFLINE_SINCE_NOT_SET) {
                State.ON
            } else {
                val now = System.currentTimeMillis()

                val minutesAgo = (now - offlineSince) / 1000 / 60

                when {
                    minutesAgo > 24 * 60 -> {
                        State.OFF
                    }

                    minutesAgo > 15 -> {
                        State.DELAYED
                    }

                    else -> {
                        State.ON
                    }
                }
            }
        }

    enum class State {
        ON, DELAYED, OFF
    }

    companion object {
        const val OFFLINE_SINCE_NOT_SET: Long = -1
    }
}
